pkg/ignore/types.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ignore
+
+import (
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// fileIgnorePattern holds the ignorable patterns
+type fileIgnorePattern struct {
+	Pattern      string
+	Paths        []string
+	RegexPattern *regexp.Regexp
+	Invert       bool
+}
+
+// Ignorable  holds the what type of ignore
+type Ignorable int
+
+const (
+	// NoIgnore marks the directory or file not to be ignored
+	NoIgnore Ignorable = iota
+	// RootDirNoIgnore marks the directory to be the root of directory walk and avoid Ignoring
+	RootDirNoIgnore
+	// TransitiveIgnore marks the directory is transitive ignore so not to be skipped
+	TransitiveIgnore
+	// CurrentIgnore marks the directory to be not having any transitive dependencies
+	// so clean be skipped from walking
+	CurrentIgnore
```

Comment:
So generally in go-style the preference is to avoid naming constants with redundant terms with the package name, e.g. here we'd have `ignore.CurrentIgnore`.

I've generally heard this referred to as "stutter" and the preference is to try and anticipate usage when naming constants to take the package name into account.  So `ignore.Current`, or `ignore.Transitive` avoid the stutter.

I don't have a good suggestion top-of-mind for `NoIgnore`, and I'm not sure why we need `RootDirNoIgnore` 🤔

Reply:
NoIgnore is more of classification not to ignore the file/dir and now I have removed RootDir constant, as that can be grouped under NoIgnore. Its more form code point of view to see that paths.